src/schema.ts

```diff
@@ -57,13 +96,184 @@ export function generateSchema(ctx: Context, fileDesc: FileDescriptorProto, sour
+        '${service.name}': {
+          options: ${serviceOptions},
+          methods: {${joinCode(methodsOptions, { on: ',' })}}
+        }
+      `);
+    }
+  });
+
+  const enumsOptions: Code[] = [];
+  (fileDesc.enumType || []).forEach((Enum) => {
+    const valuesOptions: Code[] = [];
+    Enum.value.forEach((value) => {
+      if (value.options) {
+        const valueOptions = encodedOptionsToOptions(ctx, (value.options as any)['_unknownFields']);
+        delete (value.options as any)['_unknownFields'];
+        if (valueOptions) {
+          valuesOptions.push(code`'${value.name}': ${valueOptions}`);
+        }
+      }
+    });
+
+    let enumOptions: Code | undefined;
+    if (Enum.options) {
+      enumOptions = encodedOptionsToOptions(ctx, (Enum.options as any)['_unknownFields']);
+      delete (Enum.options as any)['_unknownFields'];
+    }
+
+    if (valuesOptions.length > 0 || enumOptions) {
+      enumsOptions.push(code`
+        '${Enum.name}': {
+          options: ${enumOptions},
+          values: {${joinCode(valuesOptions, { on: ',' })}}
+        }
+      `);
+    }
+  });
+
   chunks.push(code`
     export const ${def('protoMetadata')}: ProtoMetadata = {
       fileDescriptor: ${fileDescriptorProto}.fromPartial(${descriptor}),
       references: { ${joinCode(references, { on: ',' })} },
       dependencies: [${joinCode(dependencies, { on: ',' })}],
+      ${
+        fileOptions || messagesOptions.length > 0 || servicesOptions.length > 0 || enumsOptions.length > 0
+          ? code`options: {
+          ${fileOptions ? code`options: ${fileOptions},` : ''}
+          ${messagesOptions.length > 0 ? code`messages: {${joinCode(messagesOptions, { on: ',' })}},` : ''}
+          ${servicesOptions.length > 0 ? code`services: {${joinCode(servicesOptions, { on: ',' })}},` : ''}
+          ${enumsOptions.length > 0 ? code`enums: {${joinCode(enumsOptions, { on: ',' })}}` : ''}
+        }`
+          : ''
+      }
     }
   `);
 
   return chunks;
 }
+
+function getExtensionValue(ctx: Context, extension: FieldDescriptorProto, data: Uint8Array[]): Code {
+  if (extension.type == FieldDescriptorProto_Type.TYPE_MESSAGE) {
+    const typeName = basicTypeName(ctx, extension);
+    const resultBuffer = Buffer.concat(
+      data.map((d) => {
+        // Skip length byte
+        const reader = new Reader(d);
+        reader.uint32();
+        return (reader.buf as Buffer).slice(reader.pos);
+      })
+    );
+    const result = resultBuffer.toString('base64');
+    return code`'${extension.name}': ${typeName}.decode(Buffer.from('${result}', 'base64'))`;
+  } else {
+    const reader = new Reader(data[0]);
+    let value = (reader as any)[toReaderCall(extension)]();
+    if (typeof value === 'string') {
+      value = code`"${value}"`;
+    }
+    return code`'${extension.name}': ${value}`;
+  }
+}
+
+function encodedOptionsToOptions(ctx: Context, encodedOptions: { [key: number]: Uint8Array[] }): Code | undefined {
```

Comment:
Just for my own future future/maintenance, can we add like an in-source comment along the lines of:

```
/** Takes the protoc's input of options as proto-encoded messages, and turns them into embedded-able-in-source-code representations. */
```